### ISSUE TITLE
contexts doesn't work in queryParams of detectIntent() method calling  by event name

### ISSUE THREAD

**Issue by kodekracker** [dialogflow/dialogflow-nodejs-client-v2#64]
#### Environment details

  - OS: centos7
  - Node.js version: 8.4.0
  - npm version: 5.60
  - `dialogflow` version: 0.3.0

#### Steps to reproduce
I have created a test intent named as **i_test_context** which i want to be matched by event, and this intent also takes input context named as **logged-in-user-data** .
![screenshot from 2018-04-20 19-12-55](https://user-images.githubusercontent.com/1386835/39054322-f2b19f0a-44ce-11e8-9cde-584d5dadaeff.png)

Now, I tried to call **detectIntent** by passing event details(name and language code) and i also passed the **contexts** [More about this](https://dialogflow.com/docs/reference/api-v2/rest/v2/projects.agent.sessions/detectIntent?hl=en-GB#QueryParameters.FIELDS.contexts) field in **queryParam** .
So, my payload for **detectIntent()** method looks like this:-
```json
{
  "session": "projects/dev-enterprise/agent/sessions/fa4607f4-bdbc-4645-bcf6-7b3fa6d6fc5f",
  "queryInput": {
    "event": {
      "name": "i_test_context_event",
      "languageCode": "en"
    }
  },
  "queryParams": {
    "contexts": [
      {
        "name": "projects/dev-enterprise/agent/sessions/fa4607f4-bdbc-4645-bcf6-7b3fa6d6fc5f/contexts/logged-in-user-data"
      }
    ]
  }
}
```
But, in the response i got **Default fallback intent** as matched intent. 

### Expected Result
**i_test_context** should be matched

Thanks!

**Comment by Jefftopia** [comment]
I'm also getting default intent for Event queries.

**Comment by Jefftopia** [comment]
Update: 

In my above request, `bdy.phone ` was actually `null`. When I supplied a value (which is an Entity), the query failed, falling back to `input.unknown`.

This is a bug, and it's blocking me. Can we get some attention on this? Thanks.

**Comment by kalanyuz** [comment]
Environment details
OS: macOS 10.13.6
Node.js version: 8.11.4
npm version: 6.4.0
dialogflow version: v0.6.0

I'm experiencing the exact same problem. My payload is even simpler:
`{
  4       session: sessionPath,
  5       queryInput: {
  6         event: {
  7           name: 'cant_find_user',
  8           languageCode: languageCode,
  9         } 
 10       },
 11       outputAudioConfig: {
 12         audioEncoding: `OUTPUT_AUDIO_ENCODING_LINEAR_16`,
 13         sampleRateHertz: 44100
 14       }
 15     }; 
`
But default fallback intent is returned

**Comment by valgaze** [comment]
(This is issue is closed, but for posterity)

@Jefftopia Not sure what your system was doing, but if the data wasn't available where you thought it should be, it may have needed an extra step

You need to serialize any "parameters"/data you send to DialogFlow (this not just for events, but contexts, requests data & anything else too) into an "Interface description language" (IDL)
 called **[Protbuff](https://developers.google.com/protocol-buffers)** to communicate w/ a gRPC system

Send data to DialogFlow: JSON > Protobuf
Receive data from DialogFlow: Protobuf > JSON

*This is a really well-explained example (the author wrote DialogFlow's fulfillment library): https://stackoverflow.com/questions/47583996/send-parameters-to-webhook-on-dialogflow-sdk-v2/47666256#47666256

* Handy library to hide the details: https://www.npmjs.com/package/pb-util 

* This document is handy for a variety of reasons: https://cloud.google.com/dialogflow/docs/reference/rest/v2/projects.agent.sessions/detectIntent

* Old commit: https://github.com/googleapis/nodejs-dialogflow/blob/d907f3af57438c3e88c274b100eab520b5b51062/samples/structjson.js

**tl;dr:** If this is your parameter "data":

```json
{
	"a": 1,
	"b": 2,
	"c": "bongo",
	"d": [1, 2, "bongo", 4]
}
```

Whenever you interact w/ DialogFlow it needs to be serialized/transformed to this (& vice versa):

```json

 {
 	"fields": {
 		"a": {
 			"kind": "numberValue",
 			"numberValue": 1
 		},
 		"b": {
 			"kind": "numberValue",
 			"numberValue": 2
 		},
 		"c": {
 			"kind": "stringValue",
 			"stringValue": "bongo"
 		},
 		"d": {
 			"kind": "listValue",
 			"listValue": {
 				"values": [{
 					"kind": "numberValue",
 					"numberValue": 1
 				}, {
 					"kind": "numberValue",
 					"numberValue": 2
 				}, {
 					"kind": "stringValue",
 					"stringValue": "bongo"
 				}, {
 					"kind": "numberValue",
 					"numberValue": 4
 				}]
 			}
 		}
 	}
 }
```
